File: app/src/main/java/me/pwcong/jpstart/component/fragment/JPStartTabFragment.java
package me.pwcong.jpstart.component.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;

import java.util.List;

import butterknife.BindView;
import me.pwcong.jpstart.App;
import me.pwcong.jpstart.R;
import me.pwcong.jpstart.adapter.JPStartTabPagerAdapter;
import me.pwcong.jpstart.mvp.bean.JPTab;
import me.pwcong.jpstart.mvp.presenter.BasePresenter;
import me.pwcong.jpstart.mvp.presenter.JPStartTabFragmentPresenterImpl;
import me.pwcong.jpstart.mvp.view.BaseView;

/**
 * Created by Pwcong on 2016/9/25.
 */

public class JPStartTabFragment extends BaseFragment implements BaseView.JPStartTabFragmentView {


    @BindView(R.id.tab_layout)
    TabLayout mTabLayout;
    @BindView(R.id.view_pager)
    ViewPager mViewPager;

    private BasePresenter.JPStartTabFragmentPresenter presenter;


    @Override
    protected int getViewId() {
        return R.layout.fragment_tab;
    }

    @Override
    protected void initVariable(@Nullable Bundle savedInstanceState) {

        presenter=new JPStartTabFragmentPresenterImpl(this);

        initTabLayout();
        initViewPager();

    }

    private void initTabLayout(){
        mTabLayout.setupWithViewPager(mViewPager);
    }

    private void initViewPager(){

        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                App.CURRENT_ITEM = position;
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });

    }


    @Override
    protected void doAction() {

        presenter.initJPStartTabFragment();

    }

    @Override
    public void setData(List<JPTab> data) {

        mViewPager.setAdapter(new JPStartTabPagerAdapter(getChildFragmentManager(),data));

    }

    @Override
    public void scrollViewPager(int position) {
        mViewPager.setCurrentItem(position,true);
    }
}

File: app/src/main/java/com/github/pwcong/jpstart/ui/fragment/JPStartTabFragment.kt
package com.github.pwcong.jpstart.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.google.android.material.tabs.TabLayout
import com.github.pwcong.jpstart.App
import com.github.pwcong.jpstart.adapter.JPStartTabPagerAdapter
import com.github.pwcong.jpstart.databinding.FragmentTabBinding
import com.github.pwcong.jpstart.mvp.bean.JPTab
import com.github.pwcong.jpstart.mvp.presenter.BasePresenter.JPStartTabFragmentPresenter
import com.github.pwcong.jpstart.mvp.presenter.impl.JPStartTabFragmentPresenterImpl
import com.github.pwcong.jpstart.mvp.view.BaseView.JPStartTabFragmentView

class JPStartTabFragment : BaseFragment<FragmentTabBinding>(), JPStartTabFragmentView {

    private lateinit var mTabLayout: TabLayout
    private lateinit var mViewPager: ViewPager

    private lateinit var presenter: JPStartTabFragmentPresenter

    override fun initViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): FragmentTabBinding {
        return FragmentTabBinding.inflate(inflater, container, false)
    }

    override fun initVariable(savedInstanceState: Bundle?) {
        mTabLayout = getViewBinding().tabLayout
        mViewPager = getViewBinding().viewPager

        presenter = JPStartTabFragmentPresenterImpl(this)

        initTabLayout()
        initViewPager()
    }

    private fun initTabLayout() {
        mTabLayout.setupWithViewPager(mViewPager)
    }

    private fun initViewPager() {
        mViewPager.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(
                position: Int,
                positionOffset: Float,
                positionOffsetPixels: Int
            ) {
            }

            override fun onPageSelected(position: Int) {
                App.CURRENT_ITEM = position
            }

            override fun onPageScrollStateChanged(state: Int) {
            }
        })
    }

    override fun doAction() {
        presenter.initJPStartTabFragment()
    }

    override fun scrollViewPager(position: Int) {
        mViewPager.setCurrentItem(position, true)
    }

    override fun setData(data: List<JPTab>) {
        mViewPager.adapter =
            JPStartTabPagerAdapter(childFragmentManager, data)
    }
}
